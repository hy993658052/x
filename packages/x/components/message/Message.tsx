import classnames from 'classnames';
import React from 'react';
import useXComponentConfig from '../_util/hooks/use-x-component-config';
import { useXProviderContext } from '../x-provider';
import Source from './components/Source';
import Sup, { SupProps } from './components/Sup';
import System from './components/System';
import Think from './components/Think';
import useStyle from './style';

export type MessageProps = Omit<SupProps, 'content'> & {
  prefixCls?: string;
  style?: React.CSSProperties;
  className?: string;
  rootClassName?: string;
  type?: 'sup' | 'think' | 'source' | 'system';
  content?: React.ReactNode;

  // think props
  title?: React.ReactNode;
  logo?: React.ReactNode;

  // source props
  items?: Array<{
    label: string;
    src: string;
    key?: string;
  }>;

  // system props
  divider?: boolean | string;
};

const ForwardMessage: React.FC<React.PropsWithChildren<MessageProps>> = (props) => {
  const {
    prefixCls: customizePrefixCls,
    style,
    className,
    rootClassName,
    type,
    children,
    content,
    ...restProps
  } = props;

  const { direction, getPrefixCls } = useXProviderContext();
  const prefixCls = getPrefixCls('message', customizePrefixCls);
  const contextConfig = useXComponentConfig('bubble');

  const [wrapCSSVar, hashId, cssVarCls] = useStyle(prefixCls);

  const mergedCls = classnames(
    prefixCls,
    contextConfig.className,
    className,
    rootClassName,
    hashId,
    cssVarCls,
    {
      [`${prefixCls}-rtl`]: direction === 'rtl',
    },
  );

  const ContentNode = () => {
    const childNode = content || children;
    if (type === 'sup') {
      return <Sup prefixCls={prefixCls} {...restProps} content={childNode} />;
    }
    if (type === 'think') {
      return <Think prefixCls={prefixCls} {...restProps} content={childNode} />;
    }
    if (type === 'source') {
      return <Source prefixCls={prefixCls} {...restProps} content={childNode} />;
    }
    if (type === 'system') {
      return <System prefixCls={prefixCls} {...restProps} content={childNode} />;
    }
    return childNode;
  };

  return wrapCSSVar(
    <div
      className={mergedCls}
      style={{
        ...contextConfig.style,
        ...style,
      }}
    >
      <ContentNode />
    </div>,
  );
};

type CompoundedMessage = typeof ForwardMessage & {
  Sup: typeof Sup;
  Think: typeof Think;
  Source: typeof Source;
  System: typeof System;
};

const Message = ForwardMessage as CompoundedMessage;

if (process.env.NODE_ENV !== 'production') {
  Message.displayName = 'Message';
}

Message.Sup = Sup;
Message.Think = Think;
Message.Source = Source;
Message.System = System;

export default Message;
