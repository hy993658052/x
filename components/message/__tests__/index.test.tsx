import React from 'react';

import Message from '..';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { render } from '../../../tests/utils';

describe('message', () => {
  mountTest(() => <Message content="test" />);
  rtlTest(() => <Message content="test" />);
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('Message component work', () => {
    const { container } = render(<Message content="test" />);
    const element = container.querySelector<HTMLDivElement>('.ant-message');
    expect(element).toBeTruthy();
    expect(element).toMatchSnapshot();
  });

  it('Message support content', () => {
    const { container } = render(<Message content="hello world" />);
    const element = container.querySelector<HTMLDivElement>('.ant-message');
    expect(element?.textContent).toBe('hello world');
  });

  it('Message support content with ReactNode', () => {
    const { container } = render(
      <Message content={<span className="test-content">hello world</span>} />,
    );
    const element = container.querySelector<HTMLDivElement>('.ant-message .test-content');
    expect(element?.textContent).toBe('hello world');
  });

  it('Message Should support className & style', () => {
    const { container } = render(
      <Message content="test" className="test-className" style={{ backgroundColor: 'green' }} />,
    );
    const element = container.querySelector<HTMLDivElement>('.ant-message');
    expect(element).toHaveClass('test-className');
    expect(element).toHaveStyle({ backgroundColor: 'green' });
  });

  it('Message support type markdown', () => {
    const { container } = render(<Message content="# hello world" type="markdown" />);
    const element = container.querySelector<HTMLDivElement>(
      '.ant-message .ant-message-markdown h1',
    );
    expect(element?.textContent).toBe('hello world');
  });

  it('Message support type sup', () => {
    const { container } = render(<Message content="1" type="sup" />);
    const element = container.querySelector<HTMLDivElement>('.ant-message .ant-message-sup');
    expect(element?.textContent).toBe('1');
  });

  it('Message support type think', () => {
    const { container } = render(
      <Message type="think" content={'hello world'} title={'thinking'} />,
    );
    const elementTitle = container.querySelector<HTMLDivElement>(
      '.ant-message .ant-message-think-title',
    );
    const element = container.querySelector<HTMLDivElement>(
      '.ant-message .ant-message-think-content',
    );
    expect(elementTitle?.textContent).toBe('thinking');
    expect(element?.textContent).toBe('hello world');
  });
});
