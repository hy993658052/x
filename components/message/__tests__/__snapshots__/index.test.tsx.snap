// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message Message component work 1`] = `
<div
  class="ant-message"
>
  test
</div>
`;

exports[`message rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-message ant-message-rtl"
>
  test
</div>
`;
